src/swtpm/swtpm_nvstore_linear.c

```diff
@@ -0,0 +1,476 @@
+    If the file was not at the end, any following files will be moved forward,
+    as to not leave any holes. This simplifies the allocator strategy, since it
+    allows new files to always be placed at the end.
+
+    If resize is true, this may perform a resize, so pointers into state.data or
+    state.hdr must not be kept over this function call.
+*/
+static TPM_RESULT
+SWTPM_NVRAM_Linear_RemoveFile(const char *uri,
+                              uint32_t file_nr,
+                              TPM_BOOL resize)
+{
+    TPM_RESULT rc = 0;
+    uint32_t next_offset = 0xffffffff;
+    uint32_t state_end = 0;
+    uint32_t new_len;
+    uint32_t i, cur_offset, cur_end;
+    struct nvram_linear_hdr_file *file;
+    struct nvram_linear_hdr_file old_file = state.hdr->files[file_nr];
+
+    if (old_file.offset == 0) {
+        return 0;
+    }
+
+    TPM_DEBUG("SWTPM_NVRAM_Linear_RemoveFile: removing filenr %d (resize=%d)\n",
+              file_nr, resize);
+
+    state.hdr->files[file_nr].offset = 0;
+    state.hdr->files[file_nr].data_length = 0;
+    state.hdr->files[file_nr].section_length = 0;
+
+    /* find offset of file right after the one we remove, and adjust offsets */
+    for (i = 0; i < SWTPM_NVSTORE_LINEAR_MAX_STATES; i++) {
+        file = &state.hdr->files[i];
+        if (!file->offset) {
+            continue;
+        }
+
+        cur_offset = le32toh(file->offset);
+        if (cur_offset > le32toh(old_file.offset)) {
+            if (cur_offset < next_offset) {
+                next_offset = cur_offset;
+            }
+            cur_end = cur_offset + le32toh(file->section_length);
+            if (cur_end > state_end) {
+                state_end = cur_end;
+            }
+            file->offset = htole32(cur_offset -
+                                   le32toh(old_file.section_length));
+        }
+    }
+
+    if (next_offset != 0xffffffff) {
+        TPM_DEBUG("SWTPM_NVRAM_Linear_RemoveFile: compacting\n");
+        /* if we weren't the end, compact by moving following files forward */
+        memmove(state.data + le32toh(old_file.offset),
+                state.data + next_offset,
+                state_end - next_offset);
+    }
+
+    if (resize) {
+        new_len = state.length - le32toh(old_file.section_length);
+        rc = SWTPM_NVRAM_Linear_SafeResize(uri, new_len);
+        if (rc == 0) {
+            state.length = new_len;
+        }
+    }
+
+    return rc;
+}
+
+static TPM_RESULT
+SWTPM_NVRAM_Prepare_Linear(const char *uri)
+{
+    TPM_RESULT rc = 0;
+
+    TPM_DEBUG("SWTPM_NVRAM_Prepare_Linear: uri='%s'\n", uri);
+
+    if (state.initialized) {
+        if (strcmp(state.loaded_uri, uri) == 0) {
+            /* same URI loaded, this is okay, nothing to be done */
+            return 0;
+        }
+
+        logprintf(STDERR_FILENO,
+                  "SWTPM_NVRAM_PrepareLinear: Cannot prepare twice\n");
+        return TPM_FAIL;
+    }
+
+    state.loaded_uri = malloc(strlen(uri) + 1);
```

Comment:
Is there a corresponding free() to this malloc() somewhere? It should probably be in the cleanup() function.